File: backend/app/models/escrow.py
from pydantic import BaseModel
from typing import Optional


class EscrowCreateRequest(BaseModel):
    project_id: str


class FundMilestoneRequest(BaseModel):
    escrow_object_id: str
    milestone_index: int


class ReleaseRequest(BaseModel):
    escrow_object_id: str
    milestone_index: int


class DisputeRequest(BaseModel):
    escrow_object_id: str
    milestone_index: int


class SignedTxPayload(BaseModel):
    digest: str
    signature: str


class SponsoredTxResponse(BaseModel):
    transaction_bytes: str
    digest: str

File: frontend/src/types/escrow.ts
/** Matches backend EscrowCreateRequest */
export interface EscrowCreateRequest {
  project_id: string;
}

/** Matches backend FundMilestoneRequest */
export interface FundMilestoneRequest {
  escrow_object_id: string;
  milestone_index: number;
}

/** Matches backend ReleaseRequest */
export interface ReleaseRequest {
  escrow_object_id: string;
  milestone_index: number;
}

/** Matches backend DisputeRequest */
export interface DisputeRequest {
  escrow_object_id: string;
  milestone_index: number;
}

/** Matches backend SignedTxPayload */
export interface SignedTxPayload {
  digest: string;
  signature: string;
}

/** Matches backend SponsoredTxResponse */
export interface SponsoredTxResponse {
  transaction_bytes: string;
  digest: string;
}

/** On-chain escrow object shape returned from RPC queries */
export interface OnChainEscrow {
  objectId: string;
  version: string;
  digest: string;
  owner: string;
  milestoneCount: number;
  totalAmount: number;
  status: string;
}

/** On-chain milestone data from escrow object fields */
export interface OnChainMilestone {
  index: number;
  amount: number;
  funded: boolean;
  released: boolean;
  disputed: boolean;
}

/** Result of signing and submitting a sponsored transaction */
export interface SponsoredTxResult {
  digest: string;
  effects: {
    status: { status: "success" | "failure" };
    gasUsed: {
      computationCost: string;
      storageCost: string;
      storageRebate: string;
    };
  };
}
